### PR TITLE
[REVIEW] Add support for `Series.keys()` and `DataFrame.keys()`

### LINKED ISSUE #5538: [FEA] cuDF doesn't provide `.keys()` method
**Is your feature request related to a problem? Please describe.**

cuDF doesn't provide `.keys()` method

**Describe the solution you'd like**

Implement `.keys()` method as in pandas

**Describe alternatives you've considered**

We use `.columns` instead

**Additional context**

Pandas example:
```python
import pandas as pd
data = pd.DataFrame(
data=([[1, 10, 100, 1000, 10000], [2, 20, 200, 2000, 20000]]), columns=['a', 'b', 'c', 'd', 'e'])
data.keys()
```

cuDF example:
```python
import cudf as pd

data = pd.DataFrame(
data=([[1, 10, 100, 1000, 10000], [2, 20, 200, 2000, 20000]]), columns=['a', 'b', 'c', 'd', 'e'])
data.keys() # no method
```

Traceback:
``` python
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-6-e68e32038d48> in <module>
      4 data = pd.DataFrame(
      5 data=([[1, 10, 100, 1000, 10000], [2, 20, 200, 2000, 20000]]), columns=['a', 'b', 'c', 'd', 'e'])
----> 6 data.keys() # no method

~/miniconda3/envs/rapids14/lib/python3.7/site-packages/cudf/core/dataframe.py in __getattr__(self, key)
    430                 return self[key]
    431 
--> 432         raise AttributeError("'DataFrame' object has no attribute %r" % key)
    433 
    434     @annotate("DATAFRAME_GETITEM", color="blue", domain="cudf_python")

AttributeError: 'DataFrame' object has no attribute 'keys'

```

### LINKED ISSUE #5555: [BUG] cuDF `fillna()` converts `NaN` values to `0` while Pandas does not
**Is your feature request related to a problem? Please describe.**

cuDF `fillna()` converts `NaN` values to `0` while Pandas does not

**Describe the solution you'd like**

Adjust the `fillna()` behavior to match Pandas

**Describe alternatives you've considered**
I am using the following function as a workaround for fillna() with rapids:
```python
def fillna(df1, df2):
    mask = df1.isna() & df2.notna()
    df1[mask] = df2[mask]

import cudf as pd
df1 = pd.DataFrame({'a': [1.0, None, None]})
df2 = pd.DataFrame({'a': [1.0, 2, None]})
df3 = pd.DataFrame({'a': [4.0, None, 3]})

fillna(df1['a'], df2['a'])
fillna(df1['a'], df3['a'])
df1
```

**Additional context**

Pandas behavior:
```python
import pandas as pd 

df1 = pd.DataFrame({'a': [1.0, None, None]})
df2 = pd.DataFrame({'a': [1.0, 2, None]})
df3 = pd.DataFrame({'a': [4.0, None, 3]})

df1['a'].fillna(df2['a'], inplace=True)
df1['a'].fillna(df3['a'], inplace=True)
df1
```
Output:
```
	a
0	1.0
1	2.0
2	3.0
```

cuDF behavior:
```python
import cudf as pd 

df1 = pd.DataFrame({'a': [1.0, None, None]})
df2 = pd.DataFrame({'a': [1.0, 2, None]})
df3 = pd.DataFrame({'a': [4.0, None, 3]})

df1['a'].fillna(df2['a'], inplace=True)
df1['a'].fillna(df3['a'], inplace=True)
df1
```
Output:
```
a
0	1.0
1	2.0
2	0.0
```

### LINKED ISSUE #5772: [BUG] Passing a dictionary to fillna now requires a key for every column in dataframe
After the merge of #5727 , passing a dictionary of `columns : fill value` to `fillna` now requires a key for every column in the dataframe. We should revert this to the prior behavior, to match pandas. Haven't looked closely, but I believe it's coming from this loop: https://github.com/rapidsai/cudf/blob/e63eb311089837eaa24fa871e1106ec3ce5df318/python/cudf/cudf/core/frame.py#L1318

```python
import cudf
df = cudf.DataFrame({'a': [1, 2, None], 'b': [3, None, 5]}) 
df.fillna({"a": 0, "b":3}) # works
df.fillna({"a": 0})
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-4-e0c26ce24f9d> in <module>
      2 df = cudf.DataFrame({'a': [1, 2, None], 'b': [3, None, 5]})
      3 df.fillna({"a": 0, "b":3}) # works
----> 4 df.fillna({"a": 0})

/raid/nicholasb/miniconda3/envs/rapids-tpcxbb-20200727/lib/python3.7/site-packages/cudf/core/frame.py in fillna(self, value, method, axis, inplace, limit)
   1317 
   1318         for name, col in copy_data.items():
-> 1319             copy_data[name] = copy_data[name].fillna(value[name],)
   1320         result = self._from_table(Frame(copy_data, self._index))
   1321 

KeyError: 'b'
```

```python
import pandas as pd
df = pd.DataFrame({'a': [1, 2, None], 'b': [3, None, 5]}) 
print(df.fillna({"a": 0}))
     a    b
0  1.0  3.0
1  2.0  NaN
2  0.0  5.0
```

### PR DESCRIPTION
**Pull request by galipremsagar**
- [x] Resolves #5538 
- [x] Fixes #5555
- [x] Fixes #5772

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: python/cudf/cudf/core/series.py [hunk 1]
```diff
@@ -1815,6 +1815,27 @@ def drop_duplicates(self, keep="first", inplace=False, ignore_index=False):
     def fill(self, fill_value, begin=0, end=-1, inplace=False):
         return self._fill([fill_value], begin, end, inplace)
 
+    def fillna(self, value, method=None, axis=None, inplace=False, limit=None):
+        if isinstance(value, pd.Series):
+            value = Series.from_pandas(value)
+
+        if not (is_scalar(value) or isinstance(value, (dict, Series))):
```
Comment (shwina) [inline python/cudf/cudf/core/series.py:1822]: Let's use either Pandas' `is_dict_like` or `isinstance(value, abc.Mapping` to allow for other dict-like objects.

File: python/cudf/cudf/core/column/datetime.py [hunk 2]
```diff
@@ -210,13 +210,6 @@ def fillna(self, fill_value):
             fill_value = column.as_column(fill_value, nan_as_null=False)
 
         result = libcudf.replace.replace_nulls(self, fill_value)
-        result = column.build_column(
-            result.base_data,
-            result.dtype,
-            mask=None,
-            offset=result.offset,
-            size=result.size,
-        )
```
Comment (kkraus14) [inline python/cudf/cudf/core/column/datetime.py:219]: IIRC `replace_nulls` used to not get rid of the mask. Does it now?

File: python/cudf/cudf/core/series.py [hunk 3]
```diff
@@ -1811,6 +1812,27 @@ def drop_duplicates(self, keep="first", inplace=False, ignore_index=False):
...
+            value = Series.from_pandas(value)
+
+        if not (is_scalar(value) or isinstance(value, (abc.Mapping, Series))):
+            raise TypeError(
+                f'"value" parameter must be a scalar, dict '
+                f"or Series, but you passed a "
+                f'"{type(value).__name__}"'
+            )
+
+        if isinstance(value, (dict, Series)):
```
Comment (shwina) [inline python/cudf/cudf/core/series.py:1826]: Again, we should compare with `abc.Mapping` here

File: python/cudf/cudf/core/frame.py [hunk 4]
```diff
@@ -1310,13 +1310,22 @@ def fillna(self, value, method=None, axis=None, inplace=False, limit=None):
...
 
-        if not isinstance(value, dict):
+        if isinstance(value, cudf.Series):
+            value = value.reindex(self._data.names)
+        elif isinstance(value, cudf.DataFrame):
+            if not self.index.equals(value.index):
+                value = value.reindex(self.index)
+            else:
+                value = value
+        elif not isinstance(value, dict):
```
Comment (shwina) [inline python/cudf/cudf/core/frame.py:1320]: Compare with `abc.Mapping`

File: python/cudf/cudf/core/column/datetime.py (hunk 2 above)
Comment (galipremsagar) [reply python/cudf/cudf/core/column/datetime.py:219]: It is not getting rid of mask, but if the `fill_value` has `<NA>`/`null` values, that is applied onto `self` and the `result` will have a new mask based on the locations of null values in `self` and `fill_value`

File: python/cudf/cudf/core/frame.py (hunk 4 above)
Comment (galipremsagar) [reply python/cudf/cudf/core/frame.py]: Done

File: python/cudf/cudf/core/series.py (hunk 3 above)
Comment (galipremsagar) [reply python/cudf/cudf/core/series.py]: Changed this too

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Pull request merged**